op-service/eth/blob.go

```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
 }
 
 // FromData encodes the given input data into this blob. The encoding scheme is as follows:
-//
-// First, field elements are encoded as big-endian uint256 in BLS modulus range. To avoid modulus
-// overflow, we can't use the full 32 bytes, so we write data only to the topmost 31 bytes of each.
-// TODO: we can optimize this to get a bit more data from the blobs by using the top byte
-// partially.
-//
-// The first field element encodes the length of input data as a little endian uint32 in its
-// topmost 4 (out of 31) bytes, and the first 27 bytes of the input data in its remaining 27
-// bytes.
-//
-// The remaining field elements each encode 31 bytes of the remaining input data, up until the end
-// of the input.
-//
-// TODO: version the encoding format to allow for future encoding changes
+// It reads 31 bytes and then 1 byte for three times and then 31 bytes of data from the input.
+// For all the 31 bytes of data, they are encoded into [1:32] bytes of each field element.
+// For the extra 3 bytes read from the input, they are encoded into the 1 bytes of data on the top of each field element.
+// So for each field (4 field elements), 3 bytes are encoded into 4 bytes where the highest order bit is set to 0
+// This process is repeated until all data is encoded.
+// For the first field, [1:5] bytes of the first field element will be used to encode the version and the length of the data.
 func (b *Blob) FromData(data Data) error {
 	if len(data) > MaxBlobDataSize {
 		return fmt.Errorf("data is too large for blob. len=%v", len(data))
 	}
 	b.Clear()
-	// encode 4-byte little-endian length value into topmost 4 bytes (out of 31) of first field
-	// element
-	binary.LittleEndian.PutUint32(b[1:5], uint32(len(data)))
-	// encode first 27 bytes of input data into remaining bytes of first field element
-	offset := copy(b[5:32], data)
-	// encode (up to) 31 bytes of remaining input data at a time into the subsequent field element
-	for i := 1; i < 4096; i++ {
-		offset += copy(b[i*32+1:i*32+32], data[offset:])
-		if offset == len(data) {
-			break
+
+	// first field element encodes the version and the length of the data in [1:5]
+	b[1] = EncodingVersion
+
+	// encode the length as big-endian uint24 into [2:5] bytes of the first field element
+	if len(data) < 1<<24 {
+		// Zero out any trailing data in the buffer if any
+		b[2] = byte((len(data) >> 16) & 0xFF) // Most significant byte
+		b[3] = byte((len(data) >> 8) & 0xFF)
+		b[4] = byte(len(data) & 0xFF) // Least significant byte
+	} else {
+		return fmt.Errorf("Error: length_rollup_data is too large")
+	}
```

Comment:
It's more idiomatic in go to nest only the exceptional case, e.g. (with some other simplifications): 
```suggestion
    ilen := len(data)
    if (ilen >> 24) != 0 {
        return fmt.Errorf("Error: length_rollup_data is too large")
    }
    // Zero out any trailing data in the buffer if any                                                                                                                               
    b[2] = byte(ilen >> 16)
    b[3] = byte(ilen >> 8)
    b[4] = byte(ilen)
```